marlowe/src/Language/Marlowe/Client.hs

```diff
@@ -183,6 +207,35 @@ marlowePlutusContract = do
             continueWith = autoExecuteContract theClient party
 
 
+setupMarloweParams
+    :: forall s e i o.
+       ( HasWriteTx s
+       , HasOwnPubKey s
+       , HasTxConfirmation s
+       , AsMarloweError e
+       )
+    => Ledger.ValidatorHash -> RoleOwners -> Marlowe.Contract -> Contract s e (MarloweParams, TxConstraints i o)
+setupMarloweParams rolePayoutScriptHash owners contract = mapError (review _MarloweError) $ do
+    creator <- pubKeyHash <$> ownPubKey
+    let roles = extractContractRoles contract
+    if Set.null roles
+    then do
+        let params = MarloweParams
+                { rolesCurrency = adaSymbol
+                , rolePayoutValidatorHash = rolePayoutScriptHash }
+        pure (params, mempty)
+    else do
+        let tokens = fmap (\role -> (role, 1)) $ Set.toList roles
+        cur <- mapError TokenSetupFailed $ Currency.forgeContract creator tokens
+        let rolesSymbol = Currency.currencySymbol cur
+        let giveToParty (role, pkh) = Constraints.mustPayToPubKey pkh (Val.singleton rolesSymbol role 1)
+        let distributeRoleTokens = foldMap giveToParty (AssocMap.toList owners)
```

Comment:
If I understood correctly, there is a redundancy between the roles in `owners` and the roles extracted from the contract. So I am not sure what would happen if you have a role in `owners` that is not in the contract. We certainly shouldn't transfer it, so I was going to suggest traversing the list from the contract instead. But then again, maybe what we should do is just crash. Is that what would happens?

Reply:
Yep, my mistake. Should be an error if a contract role is not in `owners`.